musicbrainz/graphql.go

```diff
@@ -0,0 +1,201 @@
+
+package musicbrainz
+
+import (
+	"database/sql"
+	"errors"
+
+	"github.com/ipfs/go-cid"
+	"github.com/meta-network/go-meta"
+)
+
+// GraphQLSchema is the GraphQL schema for the MusicBrainz META index.
+const GraphQLSchema = `
+schema {
+  query: Query
+}
+
+type Query {
+  artist(
+    name: String,
+    ipi:  String,
+    isni: String
+  ): [Artist]!
+}
+
+type Artist {
+  cid:                    String!
+  name:                   String!
+  sort_name:              String!
+  type:                   String
+  gender:                 String
+  area:                   String
+  begin_date:             String
+  end_date:               String
+  ipi:                    [String!]
+  isni:                   [String!]
+  alias:                  [String!]
+  mbid:                   String!
+  disambiguation_comment: String
+  annotation:             [String!]
+}
+`
+
+// Resolver defines GraphQL resolver functions for the schema contained in
+// the GraphQLSchema constant, retrieving data from a META store and SQLite3
+// index.
+type Resolver struct {
+	db    *sql.DB
+	store *meta.Store
+}
+
+// NewResolver returns a Resolver which retrieves data from the given META
+// store and SQLite3 index.
+func NewResolver(db *sql.DB, store *meta.Store) *Resolver {
+	return &Resolver{db, store}
+}
+
+// artistArgs are the arguments for a GraphQL artist query.
+type artistArgs struct {
+	Name *string
+	IPI  *string
+	ISNI *string
+}
+
+// Artist is a GraphQL resolver function which retrieves object IDs from the
+// SQLite3 index using either an artist name, IPI or ISNI, and loads the
+// associated META objects from the META store.
+func (g *Resolver) Artist(args artistArgs) ([]*artistResolver, error) {
+	var rows *sql.Rows
+	var err error
+	switch {
+	case args.Name != nil:
+		rows, err = g.db.Query("SELECT object_id FROM artist WHERE name = ?", *args.Name)
+	case args.IPI != nil:
+		rows, err = g.db.Query("SELECT object_id FROM artist_ipi WHERE ipi = ?", *args.IPI)
+	case args.ISNI != nil:
+		rows, err = g.db.Query("SELECT object_id FROM artist_isni WHERE isni = ?", *args.ISNI)
+	default:
+		return nil, errors.New("missing name, ipi or isni argument")
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	var resolvers []*artistResolver
+	for rows.Next() {
+		var objectID string
+		if err := rows.Scan(&objectID); err != nil {
+			return nil, err
+		}
+		cid, err := cid.Parse(objectID)
+		if err != nil {
+			return nil, err
+		}
+		obj, err := g.store.Get(cid)
+		if err != nil {
+			return nil, err
+		}
+		var artist Artist
+		if err := obj.Decode(&artist); err != nil {
+			return nil, err
+		}
+		resolvers = append(resolvers, &artistResolver{objectID, &artist})
+	}
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+	return resolvers, nil
+}
+
+// artistResolver defines GraphQL resolver functions for artist fields.
+type artistResolver struct {
+	cid    string
+	artist *Artist
+}
+
+func (a *artistResolver) Cid() string {
```

Comment:
Why are some resolver return values pointers and some not?

Reply:
If a field is optional, it must return a pointer so that `nil` represents no result (rather than a zero value which could either be no result or a result which is explicitly the zero value).